File: backend/app/api/document.py
"""O10 法规文档库接口。

- GET    /documents            文档列表（分页 + 类型/层级/关键字/状态筛选）
- GET    /documents/{id}       文档详情（元信息 + 正文预览分章展示）
- POST   /documents            管理端上传文档（txt/md/pdf/docx → 建库管道增量入库）
- PUT    /documents/{id}/status 停用（DISABLED）/ 启用（SUCCESS）——检索即时生效
- DELETE /documents/{id}       删除文档（含分块，需确认）

权限：浏览任意登录用户；上传/停用/删除 require_roles(SAFETY, ADMIN)。
"""
from __future__ import annotations

from fastapi import APIRouter, Depends, File, Form, HTTPException, Query, UploadFile

from ..core.database import get_db
from ..core.security import get_current_user, require_roles
from ..model.user import RoleId
from ..service.document_service import (
    delete_document,
    list_documents,
    get_document,
    set_document_status,
    upload_document,
)
from ..utils.response import resp

router = APIRouter(prefix="/api/v1/documents", tags=["法规文档库"])

_MANAGE = require_roles(RoleId.SAFETY, RoleId.ADMIN)


@router.get("", summary="O10 文档列表（分页 + 筛选）")
def list_docs(
    doc_type: str | None = Query(default=None, description="law/regulation/company/sop/plan/case"),
    tier: str | None = Query(default=None, description="national/province/lower"),
    keyword: str | None = Query(default=None, description="文档名模糊搜索"),
    status: str | None = Query(default=None, description="PENDING/SUCCESS/FAILED/DISABLED"),
    page: int = Query(default=1, ge=1, le=100000),
    page_size: int = Query(default=20, ge=1, le=100),
    _=Depends(get_current_user),
):
    return resp(list_documents(
        doc_type=doc_type, tier=tier, keyword=keyword, status_=status,
        page=page, page_size=page_size,
    ))


@router.get("/{doc_id}", summary="O10 文档详情（元信息 + 正文预览）")
def doc_detail(doc_id: int, _=Depends(get_current_user)):
    return resp(get_document(doc_id))


@router.post("", summary="O10 上传文档入库（SAFETY/ADMIN）")
def upload_doc(
    file: UploadFile = File(...),
    title: str | None = Form(default=None, description="文档标题（缺省用文件名）"),
    doc_type: str = Form(..., description="law/regulation/company/sop/plan/case"),
    doc_level: int | None = Form(default=None, description="行政层级 1/2/3/4"),
    region: str | None = Form(default=None, description="区域（如 四川/国家）"),
    source_url: str | None = Form(default=None, description="来源链接"),
    db: Session = Depends(get_db),
    user=Depends(_MANAGE),
):
    return resp(upload_document(
        file, title=title, doc_type=doc_type, doc_level=doc_level,
        region=region, source_url=source_url, uploader_id=user.id,
    ))


@router.put("/{doc_id}/status", summary="O10 停用/启用文档（SAFETY/ADMIN）")
def change_status(
    doc_id: int,
    payload: dict,
    user=Depends(_MANAGE),
):
    status_ = str((payload or {}).get("status") or "").strip().upper()
    return resp(set_document_status(doc_id, status_))


@router.delete("/{doc_id}", summary="O10 删除文档（含分块，SAFETY/ADMIN）")
def remove_doc(doc_id: int, user=Depends(_MANAGE)):
    return resp(delete_document(doc_id))

File: frontend/src/api/document.ts
import { request } from './request'

/** O10 文档库文档 */
export interface KnowledgeDoc {
  id: number
  name: string
  type: string
  doc_type: string
  doc_type_label: string
  doc_level: number | null
  region: string | null
  tier: 'national' | 'province' | 'lower'
  status: string
  chunk_count: number
  article_count: number
  char_count: number
  source_url: string | null
  create_time: string | null
  update_time: string | null
}

/** O10 文档详情（含正文预览） */
export interface KnowledgeDocDetail extends KnowledgeDoc {
  chapters: { chapter: string; articles: { article_no: string; content: string; seq: number }[] }[]
}

export interface DocListResult {
  page: number
  page_size: number
  total: number
  items: KnowledgeDoc[]
}

export interface DocListQuery {
  doc_type?: string
  tier?: string
  keyword?: string
  status?: string
  page?: number
  page_size?: number
}

/** 文档列表（分页 + 筛选） */
export function listDocuments(params: DocListQuery) {
  return request<DocListResult>({ url: '/documents', method: 'GET', params })
}

/** 文档详情（元信息 + 正文预览） */
export function getDocument(id: number) {
  return request<KnowledgeDocDetail>({ url: `/documents/${id}`, method: 'GET' })
}

/** 管理端上传文档（txt/md/pdf/docx → 建库管道入库） */
export function uploadDocument(payload: {
  file: File
  title?: string
  doc_type: string
  doc_level?: number
  region?: string
  source_url?: string
}) {
  const form = new FormData()
  form.append('file', payload.file)
  if (payload.title) form.append('title', payload.title)
  form.append('doc_type', payload.doc_type)
  if (payload.doc_level != null) form.append('doc_level', String(payload.doc_level))
  if (payload.region) form.append('region', payload.region)
  if (payload.source_url) form.append('source_url', payload.source_url)
  return request<{ id: number; doc_id: string; title: string; blocks: number; added_chunks: number; status: string }>({
    url: '/documents',
    method: 'POST',
    data: form,
    timeout: 300_000,
  })
}

/** 管理端停用/启用文档 */
export function setDocumentStatus(id: number, status: 'DISABLED' | 'SUCCESS') {
  return request<{ id: number; status: string; changed: boolean; deleted_chunks?: number; rebuilt_chunks?: number }>({
    url: `/documents/${id}/status`,
    method: 'PUT',
    data: { status },
  })
}

/** 管理端删除文档（含分块） */
export function deleteDocument(id: number) {
  return request<{ id: number; deleted_chunks: number }>({ url: `/documents/${id}`, method: 'DELETE' })
}
